pyee/base.py

```diff
@@ -0,0 +1,362 @@
+# -*- coding: utf-8 -*-
+
+from collections import defaultdict, OrderedDict
+from threading import Lock
+from typing import Any, Dict, Generic, List, Optional, Protocol, Tuple, TypeVar, Union
+
+from typing_extensions import Literal
+
+
+class PyeeException(Exception):
+    """An exception internal to pyee."""
+
+
+class PyeeError(PyeeException):
+    """An error internal to pyee."""
+
+
+Event = TypeVar(name="Event", contravariant=True)
+ErrorEvent = Literal["error"]
+NewListenerEvent = Literal["new_listener"]
+InternalEvent = Union[ErrorEvent, NewListenerEvent]
+
+Arg = TypeVar(name="Arg", contravariant=True)
+Kwarg = TypeVar(name="Kwarg", contravariant=True)
+
+
+class AnyHandlerP(Protocol):
+    """
+    A handler with any args or kwargs. This type is used
+    to capture the type of a handler emitted by the
+    ``new_listener`` event.
+    """
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        ...
+
+
+class HandlerP(Protocol[Event, Arg, Kwarg]):
+    """
+    An event handler.
+    """
+
+    def __call__(
+        self,
+        *args: Union[Arg, Exception, Event, InternalEvent, "AnyHandlerP"],
+        **kwargs: Kwarg,
+    ) -> Any:
+        ...
+
+
+IEvent = TypeVar(name="IEvent")
+IArg = TypeVar(name="IArg")
+IKwarg = TypeVar(name="IKwarg")
+
+
+class DecoratorP(Protocol[IEvent, IArg, IKwarg]):
+    """
+    A handler decorator function.
+    """
+
+    def __call__(
+        self, f: HandlerP[IEvent, IArg, IKwarg]
+    ) -> HandlerP[IEvent, IArg, IKwarg]:
+        ...
+
+
+class EventEmitter(Generic[Event, Arg, Kwarg]):
+    """The base event emitter class. All other event emitters inherit from
+    this class.
+
+    Most events are registered with an emitter via the ``on`` and ``once``
+    methods, and fired with the ``emit`` method. However, pyee event emitters
+    have two *special* events:
+
+    - ``new_listener``: Fires whenever a new listener is created. Listeners for
+      this event do not fire upon their own creation.
+
+    - ``error``: When emitted raises an Exception by default, behavior can be
+      overridden by attaching callback to the event.
+
+      For example::
+
+          @ee.on('error')
+          def on_error(message):
+              logging.err(message)
+
+          ee.emit('error', Exception('something blew up'))
+
+    All callbacks are handled in a synchronous, blocking manner. As in node.js,
+    raised exceptions are not automatically handled for you---you must catch
+    your own exceptions, and treat them accordingly.
+    """
+
+    def __init__(self) -> None:
+        self._events: Dict[
```

Comment:
It's tempting to break this up into multiple data structures with improved type safety:

```
self._user_events: Dict[UserEvent, "OrderedDict[HandlerP[Arg, Kwarg]. HamdlerP[Arg, Kwarg]]"]
self._error_events: "OrderedDict[HandlerP{Exception, None]], HandlerP[Exception, None]]"
self._listener_events: "OrderedDict[[HandlerP[Union[UserEvent, ErrorEvent, NewListenerEvent], HandlerP[Any, Any]]"
```

I'm concerned that I won't be able to narrow the type coming in from the type-unsafe `on` method, so I'll probably need to do something like:

```python
self._all_events: Dict[Union[UserEvent, ErrorEvent, NewListenerEvent], "OrderedDict[HandlerP[Any, Any]]"]
```

But the reverse problem is true as well - `emit_error` would need to take a `Handler[Any, Any]` if pulling from `self._all_events` and convince pyright that it's a `Handler[Exception, None]`. One thing I can do is maintain all four data structures. I could also potentially make a wrapper type:

```python

class InternalHandlerP(Generic[Event, Arg, Kwarg]):
    @property
    def event(self) -> Event: ...

    @property
    def handler(self) -> HandlerP[Arg, Kwarg]: ...

class UserHandler(InternalHandler[Event, Arg, Kwarg]):
    def __init__(self, event: UserEvent, handler: HandlerP[Arg, Kwarg]): ...
        self.event: UserEvent = event
        self.handler: HandlerP[Arg, Kwarg] = handler

class ErrorHandler(InternalHandlerP[ErrorEvent, HandlerP[Exception, None]):
    def __init__(self, handler: HandlerP[Exception, None]):
        self.event: ErrorEvent = "error"
        self.handler: HandlerP[Exception, None]

class NewListenerHandler(InternalHandlerP[NewListenerEvent, HandlerP[Union[UserEvent, ErrorEvent, NewListenerEvent, HandlerP[Any, Any], None]):
    def __init__(self, handler: HandlerP[Exception, None]):
        self.event: NewListenerEvent = "new_listener"
        self.handler: HandlerP[Any, Any] = handler

...

self._events: Dict[Union[Event, ErrorEvent, NewListenerEvent], "OrderedDict[HandlerP[Any, Any], InternalHandlerP]"]
```

This way, I should be able to do something like:

```python
for handler in ist(self._events[event].values()):
    if handler.event == "error":
        # We should be able to infer the type of handler.handler now
    elif handler.event == "new_listener":
        # Same
    else:
        # Should be a HandlerP[Arg, Kwarg]
```

I think this will work!

Reply:
These data structures didn't work out with the existing abstractions, but it's definitely a step closer and I should reference this later.